addons/web_editor/models/ir_attachment.py

```diff
@@ -84,3 +84,36 @@ def _can_bypass_rights_on_media_dialog(self, **attachment_data):
         - Non admin user uploading an unsplash image (bypass binary/url check)
         """
         return False
+
+    def write(self, vals):
+        res = super().write(vals)
+        for attachment in self:
+            if attachment.mimetype in SUPPORTED_IMAGE_MIMETYPES:
+                new_checksum = attachment.checksum
+                image_data = self.env['web_editor.image.data'].search([
+                    ['res_model', '=', attachment.res_model],
+                    ['res_id', '=', attachment.res_id],
+                    ['res_field', '=', attachment.res_field],
+                ])
+                if image_data and new_checksum != image_data.image_checksum:
+                    # If an image field is being replaced and the
+                    # `web_editor.image.data` record does not contain the
+                    # correct data, unlink it as it is obsolete. E.g an image
+                    # field is modified with the editor and then replaced from
+                    # the backend.
+                    image_data.unlink()
+        return res
+
+    def unlink(self):
+        for attachment in self:
+            if attachment.mimetype in SUPPORTED_IMAGE_MIMETYPES:
+                image_data = self.env['web_editor.image.data'].search([
+                    ['res_model', '=', attachment.res_model],
+                    ['res_id', '=', attachment.res_id],
+                    ['res_field', '=', attachment.res_field],
+                ])
+                if image_data:
+                    # If an image field is unlinked, do the same for its
+                    # corresponding image date.
+                    image_data.unlink()
+        return super().unlink()
```

Comment:
I think overriding `unlink` is not the right practice (I think it is not called if the record is deleted during app uninstall for example). Check this: https://www.odoo.com/documentation/17.0/developer/reference/backend/orm.html?highlight=ondelete#odoo.api.ondelete

Reply:
I am not sure it was the right resource you wanted to link. Indeed, in this doc, they explain the `ondelete` decorator that is used to raise an exception and not unlink some record under some circumstances. If I am not mistaken, in our case, this is not what we want: we always want to unlink the attachment but  we also want to unlink the `image.data`record if it is linked to the same image than the attachment. In fact, from what I understand of the doc you linked it also seems to me that `unlink` is called at module uninstall.